### PR TITLE
Avoid a copy in ShuffleMapTask 

### PR DESCRIPTION
**Pull request by shivaram**
Avoid a copy in ShuffleMapTask by creating an iterator that will be used by the block manager.

### CODE REVIEW

**Comment by mateiz** [comment]
Looks good, thanks!

File: core/src/main/scala/spark/scheduler/ShuffleMapTask.scala [hunk 1]
```diff
@@ -124,14 +124,18 @@ class ShuffleMapTask(
...
     for (i <- 0 until numOutputSplits) {
       val blockId = "shuffleid_" + dep.shuffleId + "_" + partition + "_" + i
-      val arr = new ArrayBuffer[Any]
       val iter = buckets(i).entrySet().iterator()
-      while (iter.hasNext()) {
-        val entry = iter.next()
-        arr += ((entry.getKey(), entry.getValue()))
+      // Create an iterator to avoid a copy. blockManager will traverse this
+      // to create an block
+      val entryIter = new Iterator[Any] {
```
Comment (rxin) [inline core/src/main/scala/spark/scheduler/ShuffleMapTask.scala:130]: This is a common enough pattern to warrant its own class. Can you pull this out and make it a class, e.g. JavaMapIteratorWrapper?

Comment (JoshRosen) [inline core/src/main/scala/spark/scheduler/ShuffleMapTask.scala:130]: It looks like this pattern of obtaining a Scala `Iterator[(_, _)]` from a `HashMap` can be expressed using [`JavaConversions.mapAsScalaMap`](http://www.scala-lang.org/api/current/index.html#scala.collection.JavaConversions$); I'm not sure if there's a performance difference, though:

``` scala
import scala.collection.JavaConversions
import java.util.HashMap

object Test {
    def main(args: Array[String]) {
        val javaMap = new HashMap[Int,Int]()
        javaMap.put(1, 2)
        javaMap.put(3, 4)
        val scalaMap = JavaConversions.mapAsScalaMap(javaMap)
        val iter : Iterator[Any] = scalaMap.toIterable.iterator
        println(iter.next())
        println(iter.next())
    }
}
```

**Comment by shivaram** [comment]
Thanks Josh for pointing this ! I don't think there is a performance concern as it just wraps the java map https://github.com/scala/scala/blob/2.9.x/src/library/scala/collection/JavaConversions.scala#L485 

Making a change to just use the JavaConversions call now...

**Comment by shivaram** [comment]
New pull request at https://github.com/mesos/spark/pull/168

File: core/src/main/scala/spark/scheduler/ShuffleMapTask.scala (hunk 1 above)
Comment (JoshRosen) [inline core/src/main/scala/spark/scheduler/ShuffleMapTask.scala:130]: Oops, `scalaMap.toIterable.iterator` should be replaced by `scalaMap.iterator`.

**Pull request merged**